File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let result = 0;
    
    const dfs = function (root) {
        if (!root) return 0;
        let right = dfs(root.right);
        let left = dfs(root.left);
        result = Math.max(result, left + right + 1);
        return Math.max(left, right) + 1;
    }
    
    dfs(root);
    
    return result - 1;
};

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        result = 0
        
        def dfs (root):
            nonlocal result
            if not root: return 0
            right, left = dfs(root.right), dfs(root.left)
            result = max(result, left + right + 1)
            return max(left, right) + 1
        
        dfs(root)
        
        return result - 1
